### ISSUE TITLE
[migration] Auditing the current DBs (Postgres and ElasticSearch) for list of datasets and data

### ISSUE THREAD

**Issue by rufuspollock** [openspending/openspending#1480]
When preparing for the migration I want to audit the current PostgreSQL DB (and ElasticSearch DB) to get list of datasets and their size (and maybe access frequency) so I can understand what needs to be migrated and what is involved.

## 🗃️ Outputs

* Spreadsheet with overview of database datasets / data tables: https://docs.google.com/spreadsheets/d/1Cno0jUkfl8ozf0qjaNhuCmJEmMaL2U65sNONqVTEEnM/edit
* Repo with results of audit: 🚧  https://github.com/openspending/audit-for-migration-2020

## Acceptance

* [x] Database is accessible to @rufuspollock and @sglavoie.
* [x] Have info on "data" tables. Specifically I want 2 "CSV" (then google sheet) tables:
  * [x] Database list: name, size, no of tables. **DONE:** Updated in `Databases` sheet in the above spreadsheet.
  * [x] (Data) Table list (for openspending only i think) with following fields `name | size | row count | owner | owner email |`
    * **Done**.
    * We also want to try and find the non-data tables e.g. the table that list all the data tables with their owners (i assume such a table exists). It will probably be named a bit differently.
    * owner is owner in terms of openspending not postgres
* [x] List of datasets and owners (from elasticsearch rather than postgres I think): 

~~NOTE: (I assume) DB has both "application" tables and "data" tables. Data tables contain the uploaded fiscal info. Applications tables are used for running the web app. I guess there should be an application table for e.g. listing the data tables (but i could be wrong - this info might be in elasticsearch for example)~~ **❌  Postgres only has data. Metadata listing datasets is in Elasticsearch**

## Tasks

* [x] Explore the database and make sure that we can retrieve all the necessary data.
* [x] Determine what, if any, calculations should be performed to analyze the content of the database.
* [x] List what details are needed as part of the auditing process. **DONE**
  * [x] Obtain basic info/stats from database(s). **DONE**
  * [x] Push to spreadsheet. **DONE**
* [x] Get clear on what needs to be done once the data is available.
  * [x] Determine where the relevant information should be stored and in which format.

Tasks as 3 July

* [x] Add key figures from current README to overview of openspending in spreadsheet. **DONE**
* [x] Pushing existing audit section with readme (but w/o access info) to https://github.com/openspending/audit-for-migration-2020. **DONE**
  * [x] Remove stuff now in spreadsheet (and put link to spreadsheet somewhere)
* [x] Add an sql file of the commands you ran. **DONE**
  * [ ] (? if quick i.e. < 2h) Bonus a python script for consolidating the output from those sql commands

**Comment by sglavoie** [comment]
### Done

* Retrieve the missing information and statistics for the databases precisely and put it in the spreadsheet. → https://docs.google.com/spreadsheets/d/1Cno0jUkfl8ozf0qjaNhuCmJEmMaL2U65sNONqVTEEnM/edit#gid=558049663&range=A1
* Add key figures to the Overview sheet.
* Push existing audit section, limited to README and list of SQL commands. → https://github.com/openspending/audit-for-migration-2020

### To do

* The ownership information is not available in the expected way (PostgreSQL or API). I will consolidate what I can from what can be seen publicly.
* (maybe) I haven't made a Python script as the outputs from the SQL commands work great. I can retrieve the desired info for a given table and export to CSV quickly without needing to reformat anything. Unless we will need to retrieve this information on a daily basis, I thought it wasn't worth spending the extra time to work on this right now. 

### Question

* _Remove stuff now in spreadsheet (and put link to spreadsheet somewhere)_ → I didn't get what you meant. Should the information it contains remain private? I have limited access to the spreadsheet while waiting for your confirmation. Some pieces of information are missing at https://github.com/openspending/audit-for-migration-2020, in the README as well as the database samples: I can add them quickly once I am clear on what needs to be there exactly.

**Comment by sglavoie** [comment]
**TL;DR**: I am blocked with finding the ownership information. Do you reckon I can reach out to Adam to get more help with debugging the issues I am facing? Else, I will need to see if someone else might be as familiar with this setup.

### Blockers

* I couldn't make more progress with finding the ownership information for each table from Elasticsearch: I was unable to log into the OpenSpending containers with the credentials I have from the OKF DevOps.
* The API at http://openspending.org/search/package only returns public information for the first 100 packages and seems to constitute a dead-end.

**Comment by rufuspollock** [comment]
@sglavoie i think you could also push some of the sample data you had as well.

> Remove stuff now in spreadsheet (and put link to spreadsheet somewhere)

What i meant was that you could put a link to the spreadsheet from https://github.com/openspending/audit-for-migration-2020 and remove the tables you have there since they duplicate spreadsheet ...

> The API at http://openspending.org/search/package only returns public information for the first 100 packages and seems to constitute a dead-end.

Why is this a dead-end? Because data is not right or b/c you can't get all the datasets? If the latter, have you asked about pagination or query parameters or look up the service that does this in openspending set of repos and seen if there are docs e.g. https://openspending.readthedocs.io/en/latest/

I also note from brief perusal that it looks like datapackage.json *and* data maybe stored in raw storage on s3 - and you could ask @akariv how one accesses / lists those.

**Comment by sglavoie** [comment]
> push some of the sample data
> could put a link to the spreadsheet

Done!

> have you asked about pagination or query parameters or look up the service that does this in openspending

Yes, I have seen `page` and `pagesize` being available as parameters in the https://github.com/openspending/os-api repo. I faced some issues when installing dependencies and also when using the Docker images available at https://hub.docker.com/u/openspending. I will reach out to Victor Nițu who kindly offered to help with this since he has worked on the project before. 
> ask akariv how one accesses / lists those.

I will do that. :+1:

**Comment by rufuspollock** [comment]
@sglavoie 

> Yes, I have seen page and pagesize being available as parameters in the https://github.com/openspending/os-api repo. I faced some issues when installing dependencies and also when using the Docker images available at https://hub.docker.com/u/openspending. I will reach out to Victor Nițu who kindly offered to help with this since he has worked on the project before.

Why would you be installing this yourself? you just need to use the docs to call the live running service!

**Comment by sglavoie** [comment]
> Why would you be installing this yourself?

@rufuspollock, I understood that's what I needed to do from Adam's earlier comment: _"You can log into any of the os containers and use rhe es API using curl or similar tool."_ The public API only return public information. Emails are not actually shown (and author names are often missing) and appear like so:

```
...
"package": {
  "author": " <Adam@not.shown>",
  ...
},
...
```

This is why I assumed I would have to be logged in to be able to see more than this. In any case, I will definitely want to chat with Victor and Adam to get clear on this setup and also on how to access the necessary APIs and storage on s3 if that's the best option.

**Comment by sglavoie** [comment]
**FIXED**: Everything was done. The table selection of what to keep is still ongoing in a separate issue #1482, but the tasks listed here are done.

**Comment by rufuspollock** [comment]
@sglavoie 👏  

Qu: have you documented in the README of that repo how you obtained the data e.g. how did you get owner info etc. If not, can you document that in the README of that repo.

**Comment by sglavoie** [comment]
@rufuspollock:

> have you documented in the README of that repo how you obtained the data

Not yet. I will do that. :+1: 

> Also why do all the emails have "not.shown"?

I do not know. This is being retrieved straight from Elasticsearch so that's the way it has been stored for some reason. A few emails are fully shown, but the vast majority of them are not.

Also, many tables do not match any owner and that helps in seeing what is being duplicated when the name of the table is almost identical but without owner. I will carefully add new filters to discard those duplicates.
